File: Dynamic_Programming/c++/15_파일합치기.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/11066
// 1

#include <iostream>
#include <algorithm>
#include <cstring>

using namespace std;

int a[501], s[501], d[501][501];

int process(int i, int j) {
    if (i == j)
        return 0;
    if (d[i][j] != -1)
        return d[i][j];
    
    int &ans = d[i][j];
    for (int k = i; k <= j - 1; k++) {
        int tmp = process(i, k) + process(k + 1, j) + s[j] - s[i - 1];
        if (ans == -1 || ans > tmp)
            ans = tmp;
    }

    return ans;
}

int main() {
    int t;
    cin >> t;

    while (t--) {
        memset(d, -1, sizeof(d));
        memset(s, 0, sizeof(s));
        int n;
        cin >> n;

        for (int i = 1; i <= n; i++) {
            cin >> a[i];
            s[i] = s[i - 1] + a[i];
        }
        cout << process(1, n) << '\n';
    }
}

File: Dynamic_Programming/python/15_파일합치기.py
# fast campus 강의
# https://www.acmicpc.net/problem/11066
# 2

def process():
    N, A = int(input()), [0] + list(map(int, input().split()))
    # S[i]는 1번부터 i번까지의 누적합
    S = [0 for _ in range(N + 1)]
    for i in range(1, N + 1):
        S[i] = S[i - 1] + A[i]
    print(S)
    # DP[i][j] = i에서 j까지 합하는데 필요한 최소 비용
    # DP[i][k] + DP[k + 1][j] + sum(array[i] ~ array[j])
    DP = [[0 for i in range(N + 1)] for i in range(N + 1)]
    for i in range(2, N + 1): # 부분파일의 길이
        for j in range(1, N + 2 - i): # 시작점
            DP[j][j + i - 1]= min(DP[j][j + k] + DP[j + k + 1][j + i - 1] for k in range(i - 1)) + (S[j + i - 1] - S[j - 1])
            for r in DP:
                for c in r:
                    print(c, end='\t')
                print()
            print()
    
    print(DP[1][N])

for _ in range(int(input())):
    process()